### DISCUSSION TITLE
How to dynamically update the proxy address of an object after it is created, and whether there is a similar method in HTTPX like using "session.proxies.update"?

### DISCUSSION THREAD

**Discussion by RavenCri** [encode/httpx#2618] [General]
Please tell me how to do it. Thank you very much. I need the help of the community.
```
import requests
session = requests.Session()
proxies = {
    "http": "http://old-proxy-ip:port",
    "https": "http://old-proxy-ip:port",
}
session.proxies.update(proxies)
new_proxies = {
    "http": "http://new-proxy-ip:port",
    "https": "http://new-proxy-ip:port",
}
session.proxies.update(new_proxies)
```

**Comment by lovelydinosaur** [comment]
We don't support that usage. Proxy setup is done when you instantiate the `Client`.
If you need a different set of proxies, then instantiate a new `Client`.

**Reply by ghost** [reply]
May i ask why not? If I, for example, wanted to write a scraping tool dynamically updating proxies would be a must. Why else would I bother using a Client and its advantages if i'm gonna be replacing it on every request?

**Reply by lovelydinosaur** [reply]
The technical advantage that the client provides is connection pooling. When you change proxies you're dropping the connection pool.

I'd suggest for that use-case, a wrapper around a pool of clients.

**Reply by ghost** [reply]
I guess so, but that would require me to create something that should've been implemented already. Requests does this.
The reason I asked is because I'm writing a tool that allows user to specify a list of proxies that'll be randomly picked.
I'd have to detect if the user has specified a proxylist, and pick a random one from that and use the wrapper. If the user hasn't specified a list, I could then use the Client.

This would make it unnecessary slow, and could've been prevented if an option was passed so that `Client.request` could use proxies.

You see what I'm getting at?

**Reply by lovelydinosaur** [reply]
> I'd suggest for that use-case, a wrapper around a pool of clients.

Actually, let me revise that. A neat way to handle this would be a custom *transport* that rotates between different proxies. You'd then pass that transport to a single client instance.

I'd be happy to help provide guidance for someone working on this.

This would actually be a *really* neat project for someone to take on and share.

> I guess so, but that would require me to create something that should've been implemented already. Requests does this.

If you want to understand fully why we have a different design decision here, and the implications, then you'd want to start by reviewing our Transport API vs. `requests` adapters, and try implementing the change to see what effect that has on lower layers of the API.

**Reply by ghost** [reply]
Hmm, interesting.
I'll see what magic I can do with a custom Transport.
It'd be amazing if you could pass a small example, since I'm very new to this libary lol

If I manage to write something working, I'll publish it (the project it'll be included in is open source anyways)

**Reply by ChrisYP** [reply]
Hi there, have you finished yet?

**Reply by ghost** [reply]
No, sorry. I dropped HTTPX in favor of python3-requests as it was just easier to use,

Sorry!

**Reply by qkxie** [reply]
Open a new client is not a good idea in some situation. Create a new client will cost about 50 ms, but in some micro-service calling requests, 50 ms is too expensive.

**Comment by Emmeline-Hawthorne** [comment]
In HTTPX, you can't update the session's proxy dictionary dynamically like in requests. Instead, you can create a new client with the updated proxy settings.
